sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,750 @@
+there are never two kernels with different required sub-group sizes in the same
+image.)
+
+There is a "reqd\_work\_group\_size" property if the image contains any kernels
+with a required work-group size.  The value of the property is a `BYTE_ARRAY`
+with the following layout:
+
+```
+<dim_count (uint32)> <dim0 (uint32)> ...
+```
+
+Where `dim_count` is the number of work group dimensions (i.e. 1, 2, or 3), and
+`dim0 ...` are the values of the dimensions from the
+`[[reqd_work_group_size()]]` attribute, in the same order as they appear in the
+attribute.
+
+**NOTE**: One may wonder why the type of the "reqd\_sub\_group\_size" property
+is not `PI_PROPERTY_TYPE_UINT32` since its value is always 32-bits.  The
+reason is that we may want to expand this property in the future to contain a
+list of required sub-group sizes.  Likewise, the "reqd\_work\_group\_size"
+property may be expanded in the future to contain a list of required work-group
+sizes.
+
+
+### Changes specific to AOT mode
+
+In AOT mode, for each AOT target specified by the `-fsycl-targets` command
+line option, DPC++ normally invokes the AOT compiler for each device IR module
+resulting from the sycl-post-link tool. For example, this is `ocloc` command
+for Intel Gen AOT target and `opencl-aot` command for the x86 AOT target with
+SPIR-V as the input, or other specific tools for the PTX target with LLVMIR
+bitcode input. This causes a problem, though, for IR modules that use optional
+features because these commands could fail if they attempt to compile IR using
+an optional feature that is not supported by the target device.  We therefore
+need some way to avoid calling these commands in these cases.
+
+The overall design is as follows.  The DPC++ installation includes a
+configuration file that has one entry for each device that it supports. Each
+entry lists the set of aspects that the device supports and a list of the
+sub-group sizes that it supports.  DPC++ then consults this configuration
+file to decide whether to invoke a particular AOT compiler on each device IR
+module, using the information from the module's "SYCL/image-requirements"
+property set.
+
+#### Device configuration file
+
+The configuration file uses a simple YAML format where each top-level key is
+a name of a device architecture. These names correspond to SYCL aspect enum
+identifiers as defined in the [TBD] API header. There are sub-keys under each
+device for the supported aspects, sub-group sizes and AOT compiler ID.  For
+example:
+
+```
+gen11_1:
+  aspects: [1, 2, 3]
+  sub-group-sizes: [8, 16]
+  aot-compiler-id: gen-spir64
+gen_icl:
+  aspects: [2, 3]
+  sub-group-sizes: [8, 16]
+  aot-compiler-id: gen-spir64
+x86_64_avx512:
+  aspects: [1, 2, 3, 9, 11]
+  sub-group-sizes: [8, 32]
+  aot-compiler-id: x86-spir64
+```
+
+The values of the aspects in this configuration file can be the numerical
+values from the `enum class aspect` enumeration or the enum identifer itself.
```

Comment:
> I think we should only allow the numerical aspect values in the configuration file, otherwise we set up a needless dependency between tools that read the configuration file and the set of aspects supported in DPC++.
...
> Although I agree that aspect names are more human readable, I think this config file will be used only by ninja programmers. These users will be happy enough if the configuration file contains only aspect numbers.

Yes, I understand there is certain inconvenience for the compiler writer, but I thought of config file as a human-editable file, and this is actually one of its main purposes. Numbers seem very error-prone to me. Anyway there would have to be some place where people need to take the arch -> number from. So I believe readability is a good reason. I'd actually prefer that only mnemonics are allowed in the config file and not the numbers.

> No code needs to change if we add a new aspect in the future.

I don't think using mnemonics in the config file is anyhow different in this regard. Not to require code changes in case of numbers, we'd have to speculatively pre-populate the arch enum with numbers corresponding to new arch. Likewise we can do this using mnemonics. Start with
```
... {
  Gen_TGL,
  Gen_future_arch1,
  Gen_future_arch2
};
```
then when arch is finalized
... {
  Gen_TGL,
  Gen_future_arch1,
  Gen_future_arch2,
  Gen_concrete_arch1 = Gen_future_arch1,
  Gen_concrete_arch2 = Gen_future_arch2
};
```
If we could find a way not to hardcode specific arch at all (like my earlier example with arch aspect templated by sequence of chars), that would be even better, IMO.

Reply:
> I don't think using mnemonics in the config file is anyhow different in this regard.

I think you are only considering the case where an aspect represents the name of an architecture.  There will be many aspects that corresponds to features or other characteristics of a device.  To give some examples, we already have aspects named `atomic64` (indicating that a device supports 64-bit atomics and `gpu` (indicating that a device is a GPU).  We are likely to add more aspects like this.  Do you want to maintain a table mapping all of these aspect names to their numbers?

> but I thought of config file as a human-editable file, and this is actually one of its main purposes.

I thought we planned to provide a tool that generates the config file somehow.  For example, I think it would be easy to provide a tool that the user must run on the device, and that tool will generate the config information from that device.  In such a case, the user doesn't ever read or modify the config file directly.

Reply:
> Do you want to maintain a table mapping all of these aspect names to their numbers?

yes. This mapping could be maintained automatically using e.g. TableGen tool, which would generate needed enums or strings or maps from a single source. 
The main question though is if we agree that aspects should be mnemonics rather than numbers.
@kbsmith-intel - your comment would be appreciated.

> I thought we planned to provide a tool that generates the config file somehow
Yes, I agree that we planned. Not sure about the point are trying to make, though. Do you mean my statement about human readability being one of the main purposes contradicts this anyhow? I don't see how then.
Why wouldn't we generate a binary then if we don't care about human readability? Beside of auto-generation we want that people can create/adjust it manually.

Reply:
> Yes, I agree that we planned. Not sure about the point are trying to make, though. Do you mean my statement about human readability being one of the main purposes contradicts this anyhow? I don't see how then.
Why wouldn't we generate a binary then if we don't care about human readability? Beside of auto-generation we want that people can create/adjust it manually.

Yes, my point was that if we provide a tool that generates the config file then humans won't be the primary way that this config file is created.  Therefore, it becomes less important for the aspects to be represented by name (vs. numeric values).

Of course, it does not make sense for the config file to be binary.  It's still important that humans can update the file if absolutely necessary.  For example, an AE might tell customer how to update the file as a workaround.  I'm just saying that human update will probably not be the primary method of updating this file.

I actually don't feel that strongly about this.  The only downside to storing the aspects as names (vs. numbers) is the engineering and maintenance cost.  If you feel this benefit is worth this extra cost, then I have no objection.

Reply:
Yes, I think this is important for the feature to be really useful. @kbsmith-intel - what would be your preference for aspect specification in the config file - numeric or symbolic (mnemonic)?